README.md
 <!-- badges: start -->
 [![R-CMD-check](https://github.com/fpahlke/simulatr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/fpahlke/simulatr/actions/workflows/R-CMD-check.yaml)
[![Codecov test coverage](https://codecov.io/gh/fpahlke/simulatr/branch/main/graph/badge.svg)](https://app.codecov.io/gh/fpahlke/simulatr?branch=main)
 <!-- badges: end -->


# simulatr: Generator for Reproducible Fake Data

This repository contains the model solution of *simulatr*, an example package for learning about software development best-practices for R packages.



NEWS.md

# simulatr 0.1.0.9002

* Usage of ggplot2 according to the recommendations in the vignette [Using ggplot2 in packages](https://ggplot2.tidyverse.org/articles/ggplot2-in-packages.html)
* Minor improvements, e.g., code style optimized

# simulatr 0.1.0.9001

* First release of simulatr


tests/testthat/test_f_utilities.R



test_that("For invalid input arguments 'createSeed' throws meaningful exceptions", {
    expect_error(createSeed(numberOfValues = 0), 
        "Assertion on 'numberOfValues' failed: Element 1 is not >= 1.")
    expect_error(createSeed(numberOfValues = 1, minValue = -1), 
        "Assertion on 'minValue' failed: Element 1 is not >= 1.")
    expect_error(createSeed(numberOfValues = 1, maxValue = 0), 
        "Assertion on 'maxValue' failed: Element 1 is not >= 1.")
    expect_error(createSeed(numberOfValues = 1, minValue = 101, maxValue = 100), 
        "Assertion on 'minValue < maxValue' failed: Must be TRUE.")
})

test_that("The results of 'createSeed' depend on the input arguments as expected", {
    expect_length(createSeed(), 1)
    expect_length(createSeed(numberOfValues = 9), 9)
    
    seed <- createSeed(numberOfValues = 100, minValue = 1000, maxValue = 5000)
    expect_true(all(seed >= 1000))
    expect_true(all(seed <= 5000))
})

test_that("'createSeed' returns valid seed although network connection is missing", {
    expect_warning(seed <- createSeed(
        minValue = 100000000, maxValue = 999999999,
        test_exception = "network error"
    ), "Failed to receive new seed from www.random.org: network error")
    
    expect_type(seed, "integer")
})

tests/testthat/test_f_assertions.R


test_that("'assertPackageIsInstalled' works as expected", {
    expect_no_error(assertPackageIsInstalled("stats"))
    expect_error(assertPackageIsInstalled("notExistingPackage"),
        "Package 'notExistingPackage' is needed for this function to work")
})


tests/testthat/test_f_simluate_continuous_data.R



test_that("For invalid input arguments 'getSimulatedTwoArmMeans' throws meaningful exceptions", {
    expect_error(
        getSimulatedTwoArmMeans(),
        'argument "n1" is missing, with no default'
    )
    expect_error(
        getSimulatedTwoArmMeans(n1 = 50),
        'argument "n2" is missing, with no default'
    )
    expect_error(
        getSimulatedTwoArmMeans(n1 = 50, n2 = 50),
        'argument "mean1" is missing, with no default'
    )
    expect_error(
        getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5),
        'argument "mean2" is missing, with no default'
    )
    expect_error(
        getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7),
        'argument "sd1" is missing, with no default'
    )
    expect_error(
        getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3),
        'argument "sd2" is missing, with no default'
    )
    expect_error(
        getSimulatedTwoArmMeans(n1 = 0, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4),
        "Assertion on 'n1' failed: Element 1 is not >= 1."
    )
    expect_error(
        getSimulatedTwoArmMeans(n1 = 50, n2 = -5, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4),
        "Assertion on 'n2' failed: Element 1 is not >= 1."
    )
    expect_error(
        getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4, seed = 1.5),
        "Assertion on 'seed' failed: Must be of type 'single integerish value', not 'double'."
    )
    expect_error(
        getSimulatedTwoArmMeans(
            n1 = 50, n2 = 50, mean1 = 5, mean2 = 7,
            sd1 = 3, sd2 = 4, seed = 123, alternative = "one"
        ),
        '\'arg\' should be one of "two.sided", "less", "greater"'
    )
})

test_that("Default display of 'getSimulatedTwoArmMeans' returns an output", {
    simResult <- getSimulatedTwoArmMeans(
        n1 = 10, n2 = 10, mean1 = 5, mean2 = 7,
        sd1 = 3, sd2 = 4, seed = 232323
    )
    
    expect_output(showDefault.SimulationResult(simResult))
})

test_that("With defined seed 'getSimulatedTwoArmMeans' returns reproducible results", {
    simResult <- getSimulatedTwoArmMeans(
        n1 = 10, n2 = 10, mean1 = 5, mean2 = 7,
        sd1 = 3, sd2 = 4, seed = 232323
    )

    expect_equal(simResult$n1, 10)
    expect_equal(simResult$n2, 10)
    expect_equal(simResult$allocationRatio, 1)
    expect_equal(simResult$mean1, 5)
    expect_equal(simResult$mean2, 7)
    expect_equal(simResult$sd1, 3)
    expect_equal(simResult$sd2, 4)
    expect_equal(simResult$meanSim1, 5.8835095, tolerance = 1e-07)
    expect_equal(simResult$meanSim2, 5.7358547, tolerance = 1e-07)
    expect_equal(simResult$sdSim1, 2.9701436, tolerance = 1e-07)
    expect_equal(simResult$sdSim2, 3.4836918, tolerance = 1e-07)
    expect_equal(simResult$seed, 232323)
    expect_equal(simResult$statistic, 0.10199383, tolerance = 1e-07)
    expect_equal(simResult$df, 17.560814, tolerance = 1e-07)
    expect_equal(simResult$pValue, 0.91991717, tolerance = 1e-07)
    expect_equal(simResult$confInt, c(-2.899276, 3.1945855), tolerance = 1e-07)
    expect_equal(simResult$confLevel, 0.95, tolerance = 1e-07)
    expect_equal(simResult$alternative, "two.sided")
    expect_equal(simResult$method, "Welch Two Sample t-test")
    expect_equal(simResult$data$group, factor(c(
        1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
        2, 2, 2, 2, 2, 2, 2, 2, 2, 2
    )))
    expect_equal(simResult$data$values, c(
        5.6444129, 7.0815453, 8.5169894, 9.5635767, 5.4989867,
        8.5665747, 3.1804153, 4.3349261, 6.845017, -0.39734943, 8.8410798, 2.3826052,
        4.3751252, 8.2548479, 2.8344348, 12.083721, 7.337037, 0.48579194, 4.6572878,
        6.1066166
    ), tolerance = 1e-07)
})

test_that("With undefined seed 'getSimulatedTwoArmMeans' returns valid results", {
    simResult <- getSimulatedTwoArmMeans(
        n1 = 10, n2 = 10, mean1 = 5, mean2 = 7,
        sd1 = 3, sd2 = 4
    )

    expect_s3_class(simResult, "SimulationResult")
    expect_type(simResult$data, "list")
    expect_equal(names(simResult$data), c("group", "values"))
    expect_type(simResult$seed, "integer")
})

test_that("'getSimulatedTwoArmMeans' result can be printed", {
    simResult <- getSimulatedTwoArmMeans(
        n1 = 10, n2 = 10, mean1 = 5, mean2 = 7,
        sd1 = 3, sd2 = 4, seed = 232323
    )

    expect_output(print(simResult))
})

test_that("'getSimulatedTwoArmMeans' result plot layers match expectations", {
    simResult <- getSimulatedTwoArmMeans(
        n1 = 10, n2 = 10, mean1 = 5, mean2 = 7,
        sd1 = 3, sd2 = 4, seed = 232323
    )

    p <- plot(simResult)
    expect_s3_class(p$layers[[1]], "ggproto")
    expect_s3_class(p$layers[[1]]$geom, "GeomBoxplot")
    expect_s3_class(p$layers[[1]]$stat, "StatBoxplot")
})

test_that("'getSimulatedTwoArmMeans' result x axis is labelled 'Simulated Values'", {
    simResult <- getSimulatedTwoArmMeans(
        n1 = 10, n2 = 10, mean1 = 5, mean2 = 7,
        sd1 = 3, sd2 = 4, seed = 232323
    )
    
    p <- plot(simResult)
    expect_identical(p$labels$x, "Group")
})

test_that("'getSimulatedTwoArmMeans' result y axis is labelled 'Simulated Values'", {
    simResult <- getSimulatedTwoArmMeans(
        n1 = 10, n2 = 10, mean1 = 5, mean2 = 7,
        sd1 = 3, sd2 = 4, seed = 232323
    )

    p <- plot(simResult)
    expect_identical(p$labels$y, "Simulated Values")
})

test_that("With undefined seed 'getSimulatedTwoArmMeans' returns valid results although network connection is missing", {
    expect_warning(simResult <- getSimulatedTwoArmMeans(
        n1 = 10, n2 = 10, mean1 = 5, mean2 = 7,
        sd1 = 3, sd2 = 4,  
        test_exception = "network error"
    ), "Failed to receive new seed from www.random.org: network error")

    expect_type(simResult$seed, "integer")
})




R/f_utilities.R

#'
#' @title
#' Create Seed
#'
#' @description
#' Returns one or more true random numbers which can be used as seed.
#'
#' @param numberOfValues a single integer value. Number of seeds to create, default is \code{1}.
#' @param minValue a single integer value. The minimum value that a seed can have, default is \code{1000000}.
#' @param maxValue a single integer value. The maximum value that a seed can have, default is \code{9999999}.
#' @param ... optional arguments.
#'
#' @details
#' RANDOM.ORG offers true random numbers to anyone on the Internet.
#' The randomness comes from atmospheric noise, which for many purposes
#' is better than the pseudo-random number algorithms typically used
#' in computer programs. For more information see \url{https://www.random.org}.
#'
#' @seealso \link{getSimulatedTwoArmMeans}
#'
#' @return an integer value or vector containing one or more seeds.
#'
#' @export
#'
createSeed <- function(numberOfValues = 1, minValue = 1000000, maxValue = 9999999, ...) {
    assertPackageIsInstalled("httr")
    assertPackageIsInstalled("glue")
    checkmate::assertInt(numberOfValues, lower = 1, upper = 1000)
    checkmate::assertInt(minValue, lower = 1, upper = 100000000)
    checkmate::assertInt(maxValue, lower = 1, upper = 999999999)
    checkmate::assertTRUE(minValue < maxValue)
    tryCatch(
        {
            args <- list(...)
            if (length(args) > 0 && !is.null(args[["test_exception"]])) {
                stop(args[["test_exception"]])
            }
            
            minValue <- as.integer(minValue)
            maxValue <- as.integer(maxValue)
            url <- glue::glue(paste0(
                "https://www.random.org/integers/",
                "?num={numberOfValues}",
                "&min={minValue}",
                "&max={maxValue}",
                "&col=1",
                "&base=10",
                "&format=plain",
                "&rnd=new"
            ))

            response <- httr::GET(url)
            response <- httr::content(response)
            response <- trimws(response)
            response <- strsplit(response, "\n")
            response <- unlist(response)
            return(as.integer(response))
        },
        error = function(e) {
            warning("Failed to receive new seed from www.random.org: ", e$message)
            seed <- as.integer(Sys.time())
            while (seed > maxValue) {
                seedText <- as.character(seed)
                n <- nchar(seedText)
                seed <- as.integer(substring(seedText, max(2, n - 8), n))
            }
            return(seed)
        }
    )
}


R/f_simulate_continuous_data.R

#'
#' @title
#' Get Simluated Two Arm Means
#'
#' @description
#' Simluates and returns a normally distributed continuous dataset for two groups.
#'
#' @param n1 a single integer value. The sample size of group 1.
#' @param n2 a single integer value. The sample size of group 2.
#' @param mean1 a single numeric value. The assumed mean of group 1.
#' @param mean2 a single numeric value. The assumed mean of group 2.
#' @param sd1 a single numeric value. The assumed standard deviation of group 1.
#' @param sd2 a single numeric value. The assumed standard deviation of group 2.
#' @param seed a single integer value. The start value for generating pseudo-random numbers.
#' @param alternative a character string specifying the alternative hypothesis,
#'        must be one of "two.sided" (default), "greater" or "less".
#' @param ... ensures that all arguments (starting from the "...") are to be named.
#'
#' @details
#' If the \code{seed} is not specified (NA) the function \link{createSeed} will be used to create
#' a true random number as seed.
#'
#' See \url{https://cran.r-project.org/package=simstudy} for a comprehensive R package for study data simulation.
#'
#' @seealso \link{createSeed}
#' 
#' @return a list with all arguments and results; the ouput is defined as a class with name 'SimulationResult'.
#'
#' @examples
#' getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4, seed = 123)
#'
#' @export
#'
getSimulatedTwoArmMeans <- function(
        n1, 
        n2, 
        mean1, 
        mean2, 
        sd1, 
        sd2, 
        ...,
        seed = NA_integer_,
        alternative = c("two.sided", "less", "greater")) {
    
    # use assertions to check all input arguments
    checkmate::assertInt(n1, lower = 1)
    checkmate::assertInt(n2, lower = 1)
    checkmate::assertNumber(mean1)
    checkmate::assertNumber(mean2)
    checkmate::assertNumber(sd1)
    checkmate::assertNumber(sd2)
    checkmate::assertInt(seed, na.ok = TRUE)

    # create a new seed if it is NA, i.e., it is undefined
    if (is.na(seed)) {
        seed <- createSeed(...)
    }

    # specify seed
    if (!is.na(seed)) {
        set.seed(seed)
    }

    # create normal distributed random data for the two groups
    values1 <- rnorm(n = n1, mean = mean1, sd = sd1)
    values2 <- rnorm(n = n2, mean = mean2, sd = sd2)

    # save the fake data to a data frame in long format
    data <- data.frame(
        group = as.factor(c(rep(1, n1), rep(2, n2))),
        values = c(values1, values2)
    )

    # calcualate t-test result
    testResult <- t.test(values ~ group, data = data, alternative = match.arg(alternative))

    # put all arguments and results to a list
    result <- list(
        n1 = n1,
        n2 = n2,
        allocationRatio = n1 / n2,
        mean1 = mean1,
        mean2 = mean2,
        sd1 = sd1,
        sd2 = sd2,
        meanSim1 = mean(values1),
        meanSim2 = mean(values2),
        sdSim1 = sd(values1),
        sdSim2 = sd(values2),
        seed = seed,
        statistic = testResult$statistic[[1]],
        df = testResult$parameter[[1]],
        pValue = testResult$p.value,
        confInt = as.vector(testResult$conf.int),
        confLevel = attr(testResult$conf.int, "conf.level"),
        alternative = testResult$alternative,
        method = testResult$method,
        data = data
    )

    # define that the result list is a class with name 'SimulationResult'
    result <- structure(result, class = "SimulationResult")

    return(result)
}

#'
#' @title
#' Print Simulation Result
#'
#' @description
#' Generic function to print a \code{SimulationResult} object.
#'
#' @param x a \code{SimulationResult} object to print.
#' @param ... further arguments passed to or from other methods.
#' 
#' @examples
#' x <- getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4, seed = 123)
#' print(x)
#'
#' @export
#'
print.SimulationResult <- function(x, ...) {
    assertPackageIsInstalled("dplyr")
    checkmate::checkClass(x, "SimulationResult")

    # hide the output attr(,"class")
    attributes(x)[["class"]] <- NULL
    
    # format the confidence interval output
    x$confInt <- paste0("[", paste0(x$confInt, collapse = ", "), "]")
    
    # show the fake data as tibble
    x$data <- dplyr::tibble(x$data)
    
    print(x, ...)
}

#'
#' @title
#' Show Simulation Result
#'
#' @description
#' Generic function to show a \code{SimulationResult} object.
#'
#' @param x a \code{SimulationResult} object to show.
#' @param ... further arguments passed to or from other methods.
#'
#' @examples
#' getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4, seed = 123)
#' 
#' @export
#'
showDefault.SimulationResult <- function(x, ...) {
    print(x = x, ...)
}

#'
#' @title
#' Plot Simulation Result
#'
#' @description
#' Generic function to plot a \code{SimulationResult} object.
#'
#' @param x a \code{SimulationResult} object to plot.
#' @param main an overall title for the plot.
#' @param xlab a title for the x axis.
#' @param ylab a title for the y axis.
#' @param ... ensures that all arguments (starting from the "...") are to be named.
#'
#' @details
#' Uses ggplot2 to create the plot.
#'
#' @return
#' A ggplot2 object.
#'
#' @examples
#' x <- getSimulatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4, seed = 123)
#' if (require(ggplot2)) plot(x)
#'
#' @importFrom rlang .data
#' 
#' @export
#'
plot.SimulationResult <- function(
        x, 
        ..., 
        main = "Continuous Fake Data", 
        xlab = "Group", 
        ylab = "Simulated Values") {
        
    assertPackageIsInstalled("ggplot2")
    checkmate::checkClass(x, "SimulationResult")

    # see vignette("ggplot2-in-packages")
    data <- x$data
    p <- ggplot2::ggplot(
        data = data,
        mapping = ggplot2::aes(x = .data$group, y = .data$values)
    )
    p <- p + ggplot2::geom_boxplot(data = data, mapping = ggplot2::aes(fill = .data$group))
    p <- p + ggplot2::geom_point(
        colour = "#0e414e", shape = 20,
        position = ggplot2::position_jitter(width = .1),
        size = 2
    )
    p <- p + ggplot2::stat_summary(
        fun = "mean", geom = "point",
        shape = 21, position = ggplot2::position_dodge(.75), size = 4, fill = "white",
        colour = "black", show.legend = FALSE
    )
    p <- p + ggplot2::theme_bw()
    p <- p + ggplot2::theme(
        panel.border = ggplot2::element_blank(),
        axis.line = ggplot2::element_line(colour = "black")
    )
    p <- p + ggplot2::ggtitle(main)
    p <- p + ggplot2::xlab(xlab)
    p <- p + ggplot2::ylab(ylab)
    p <- p + ggplot2::theme(legend.position = "none")
    return(p)
}


R/f_simluate_continuous_data.R

#' 
#' @title 
#' Get Simluated Two Arm Means
#' 
#' @description 
#' Simluates and returns a normally distributed continuous dataset for two groups.
#' 
#' @param n1 a single integer value. The sample size of group 1.
#' @param n2 a single integer value. The sample size of group 2.
#' @param mean1 a single numeric value. The assumed mean of group 1.
#' @param mean2 a single numeric value. The assumed mean of group 2.
#' @param sd1 a single numeric value. The assumed standard deviation of group 1.
#' @param sd2 a single numeric value. The assumed standard deviation of group 2.
#' @param seed a single integer value. The start value for generating pseudo-random numbers.
#' @param alternative a character string specifying the alternative hypothesis, 
#'        must be one of "two.sided" (default), "greater" or "less".
#' @param ... ensures that all arguments (starting from the "...") are to be named.
#' 
#' @details 
#' If the \code{seed} is not specified (NA) the function \link{createSeed} will be used to create 
#' a true random number as seed.
#' 
#' See \url{https://cran.r-project.org/package=simstudy} for a comprehensive R package for study data simulation.
#' 
#' @seealso \link{createSeed}
#' 
#' @examples 
#' getSimluatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4, seed = 123)
#' 
#' @export 
#'
getSimluatedTwoArmMeans <- function(n1, n2, mean1, mean2, sd1, sd2, ..., 
    seed = NA_integer_, 
    alternative = c("two.sided", "less", "greater")) {
    
    checkmate::assertInt(n1, lower = 1)
    checkmate::assertInt(n2, lower = 1)
    checkmate::assertNumber(mean1)
    checkmate::assertNumber(mean2)
    checkmate::assertNumber(sd1)
    checkmate::assertNumber(sd2)
    checkmate::assertInt(seed, na.ok = TRUE)
    
    if (is.na(seed)) {
        seed <- createSeed()
    }
    
    if (!is.na(seed)) {
        set.seed(seed)
    }
    
    values1 <- rnorm(n = n1, mean = mean1, sd = sd1)
    values2 <- rnorm(n = n2, mean = mean2, sd = sd2)
    
    data <- data.frame(
        group = as.factor(c(rep(1, n1), rep(2, n2))),
        values = c(values1, values2)
    )
    
    testResult <- t.test(values ~ group, data = data, alternative = match.arg(alternative))
    
    result <- list(
        n1 = n1,
        n2 = n2,
        allocationRatio = n1 / n2,
        mean1 = mean1,
        mean2 = mean2,
        sd1 = sd1,
        sd2 = sd2,
        meanSim1 = mean(values1),
        meanSim2 = mean(values2),
        sdSim1 = sd(values1),
        sdSim2 = sd(values2),
        seed = seed,
        statistic = testResult$statistic[[1]],
        df = testResult$parameter[[1]],
        pValue = testResult$p.value,
        confInt = as.vector(testResult$conf.int),
        confLevel = attr(testResult$conf.int,"conf.level"),
        alternative = testResult$alternative,
        method = testResult$method,
        data = data
    )
    
    result <- structure(result, class = "SimulationResult")
    
    return(result)
}

#'
#' @title 
#' Print Simulation Result
#' 
#' @description 
#' Generic function to print a \code{SimulationResult} object.
#' 
#' @param x a \code{SimulationResult} object to print.
#' @param ... further arguments passed to or from other methods.
#' 
#' @export 
#' 
print.SimulationResult <- function(x, ...) {
    assertPackageIsInstalled("dplyr")
    attributes(x)[["class"]] <- NULL
    x$confInt <- paste0("[", paste0(x$confInt, collapse = ", "), "]")
    x$data <- dplyr::tibble(x$data)
    print(x, ...)
}

#'
#' @title 
#' Plot Simulation Result
#' 
#' @description 
#' Generic function to plot a \code{SimulationResult} object.
#' 
#' @param x a \code{SimulationResult} object to plot.
#' @param main an overall title for the plot.
#' @param xlab a title for the x axis.
#' @param ylab a title for the y axis.
#' @param ... ensures that all arguments (starting from the "...") are to be named.
#' 
#' @details 
#' Uses ggplot2 to create the plot.
#' 
#' @return 
#' A ggplot2 object.
#' 
#' @examples 
#' x <- getSimluatedTwoArmMeans(n1 = 50, n2 = 50, mean1 = 5, mean2 = 7, sd1 = 3, sd2 = 4, seed = 123)
#' if (require(ggplot2)) plot(x)
#' 
#' @export 
#' 
plot.SimulationResult <- function(x, ..., main = "Continuous Fake Data", xlab = "Group", ylab = "Simulated Values") {
    assertPackageIsInstalled("ggplot2")
    data <- x$data
    p <- ggplot2::ggplot(
        data = x$data,
        ggplot2::aes_string(x = 'group', y = 'values')
    )
    p <- p + ggplot2::geom_boxplot(ggplot2::aes_string(fill = 'group'))
    p <- p + ggplot2::geom_point(
        colour = "#0e414e", shape = 20,
        position = ggplot2::position_jitter(width = .1),
        size = 2
    )
    p <- p + ggplot2::stat_summary(
        fun = "mean", geom = "point",
        shape = 21, position = ggplot2::position_dodge(.75), size = 4, fill = "white",
        colour = "black", show.legend = FALSE
    )
    p <- p + ggplot2::theme_bw()
    p <- p + ggplot2::theme(
        panel.border = ggplot2::element_blank(),
        axis.line = ggplot2::element_line(colour = "black")
    )
    p <- p + ggplot2::ggtitle(main)
    p <- p + ggplot2::xlab(xlab)
    p <- p + ggplot2::ylab(ylab)
	p <- p + ggplot2::theme(legend.position = "none")
    return(p)
}

